### PR TITLE
Implement RNN decoding; feed output t-1 as input t

### PR DESCRIPTION
**Pull request by israelg99**
Adds an optional argument `output_length` to the `Recurrent` layer, if `output_length` is greater than the input sequence, then after processing the input, the output of the RNN from timestep `t-1` will be used as input at timestep `t` to generate the remaining output sequence.

```
input_seq = Input((16, 32))

# Output: `(batch, 32)`
encoder = GRU(32)(input_seq)

# Output: `(batch, 1, 32)`
# Can repeat more than once.
input_decoder = RepeatVector(1)(encoder)

# Output: `(batch, 16, 32)`
# Use the encoder's output as the initial input for the decoder.
# Then the decoder feeds its output from timestep `t-1` as input at timestep `t` to generate the rest of the output sequence.
decoder = GRU(32, unroll=True, output_length=16)(input_decoder)
```

This feature is necessary for a Seq2Seq model, especially if we want to have a seq2seq example script in the `examples/` folder as @fchollet mentioned here - #5559 (I'm on it).

I tested it on latest TF and theano and it works as expected.

### CODE REVIEW

**Comment by fchollet** [comment]
Any volunteers to review this PR?

**Comment by farizrahman4u** [comment]
With the right approach, you should be able to decode even if the RNN is not unrolled.

**Comment by bstriner** [comment]
It seems like this could be implemented with current functionality if you had two separate RNNs for encoding and generating with step functions that share variables. They could actually use the same step function with permuted inputs.

If you have different lengths within a batch, so the generator takes over at different points, I would just put a switch in the step function and use a single RNN. If the current depth is less than the length, use the input, if the current depth is more than the length, use the output.

Either way, this should be straightforward to do with symbolic loops. Can you explain more about what this is trying to accomplish? What model are you trying to reproduce?

Cheers!

**Comment by israelg99** [comment]
@bstriner with the code example I provided above I'm looking to reproduce a Seq2Seq model:
seq2seq

If you think it's possible to implement it with current functionality, can you please provide a code example (seq2seq)? :)

@farizrahman4u I'm looking into it!

**Comment by bstriner** [comment]
@israelg99 simple enough and a good concrete example. Do you have a particular paper in mind? There are a lot of ways to do Seq2Seq models. A few questions:

* Are the inputs one-hot words?
* Are the outputs a softmax over output words? What is the "W" in the picture? I assume that is some sort of start output symbol. 
* Are each input and output sequence the same length or not? How would you handle padding?
* You could build a version where the ht0->ht1 function is shared between the encoder and the decoder, or not. I could imagine advantages and disadvantages to both.

The easiest option is to pad each sequence out with null characters so you can easily batch things. The better option is to use some masks.

Here is the architecture:
* input x: `[input sequence, output sequence, padding]`
* input mask: `[1 for _ in inputs, 2 for _ in outputs, 0 for _ in padding]`

Recurrent unit inputs:
* input x
* input mask
* last predicted x
* last hidden state

Recurrent unit outputs:
* hidden state
* softmax over outputs
* sample over softmax

Logic of unit is to switch based on the mask whether it is using input x or last predicted x. Unit always calculates a next hidden state, a softmax and a predicted x. The predicted x is just ignored unless it is an output step. Just make heavy use of equals and switch.

For the loss, you would just do categorical crossentropy using that softmax masked for where mask=2 or whatever.

For testing, you just pass input sequences padded with all output mask.

Easy enough to build something like that in Keras proper, but the unit is going to be custom. Something like @farizrahman4u recurrentshop might be a good way to build the unit out of something reusable.

Cheers

**Comment by bstriner** [comment]
@israelg99 whipped up a bare-bones S2S model here: https://github.com/bstriner/keras-seq2seq

Heavily undocumented but it should be a good starting point. Works as described above.

It takes an input sequence and an output sequence. It trains to minimize categorical cross entropy of predicted output sequence. In test mode, each predicted output is fed back into the network.

To handle sequences of different lengths, input and output sequence are concatenated into a single matrix and a mask indicating input, output, and padding is created as a separate matrix.

It trains on shakespeare bigrams. Input is a sequence of characters from word 1 and it predicts word 2 character-by-character.

> cry: [they]
> denmark: [antonio]
> winds: [wife]
> you: [antony]
> thou: [most]

You can try stochastic generation or just simple argmax. I tested in theano but it should hopefully work on tf as well.

Cheers

**Comment by fchollet** [comment]
What's the status on this functionality? I see that the PR wasn't formally reviewed and that contributor opinions seem to indicate the functionality might not be needed.

@farizrahman4u @bstriner: is that the consensus? Should I close the PR?

**Comment by bstriner** [comment]
@fchollet No response from OP in 3 months so probably safe to close it. Something is needed but I'm not sure this is the right approach. I don't think making Seq2Seq models easy is going to be a single PR. I also don't want to do something that only works with unrolled models.

@farizrahman4u any thoughts on getting S2S to look good in Keras?

@israelg99 any progress on your project or any code to share?

The biggest problem is if you want to pass all of your weights through non-sequences then you can't leverage a lot of existing code. All of the code within one pass of the loop needs to be within a single layer. If you want to build a S2S model correctly you have to replicate the weights and logic of an LSTM within a custom layer that loops back the sampled output. No easy way to utilize sublayers (unless you go around passing weights through non-sequences).

Some initial issues. Please correct me if there is just some function that I don't know about.
* Need multiple inputs, some of which are 3d (`sequences`) and some of which are 2d (`non_sequences`) and multiple outputs. I think some Keras examples just concatenate or replicate the inputs, which isn't ideal, especially if they might be different dtypes.
* Passing trainable weights as non-sequences means optimization should work better but means you can't utilize Keras layers. Something like recurrentshop is maybe a good approach but that code might need some work. It's a pain to constantly build LSTM units, etc. inside the layers. Has anyone actually tested how much of an impact non-sequences make?
* A few missing functions from the backend are required for creating `outputs_info` efficiently. Need to be able to dynamically generate initial values from input shapes without making unnecessary variables. `T.ones` is not in the backend and using `T.ones_like` can be awkward. If we add `T.ones` to the backend I'm not sure what we would name it to not conflict with the existing `K.ones` (which makes a variable).
* Specifically for sampling and looping back output, I think we need more RNG functions in the backend.
* You need the target sequence as you are unrolling the predictions during training so you don't want to also pass that as `targets`. Using dummy data as the `targets` you pass to train seem to be the best idea. Always use the same loss `lambda yt, yp: yp`. Can add that loss to `keras` with some documentation of why you would do that.

Whether we make those changes in `keras` or `keras-contrib` or `recurrentshop` is fine with me. Several different PRs.

If you're willing to write a custom layer as above, you can do everything just fine in Keras as is, but it looks kinda ugly.

Cheers

**Comment by fchollet** [comment]
Thanks for the comments.
